src/server/game/Movement/MotionMaster.cpp

```diff
@@ -1141,20 +1138,16 @@ void MotionMaster::DirectAdd(MovementGenerator* movement, MovementSlot slot/* =
                         return a->Priority == movement->Priority;
                     });
 
-                    if (itr != _generators.end())
-                    {
-                        MovementGenerator* pointer = *itr;
-                        _generators.erase(pointer);
-                        Delete(pointer, false, false);
-                    }
+                    insertPosition = itr;
                 }
             }
             else
                 _defaultGenerator->Deactivate(_owner);
 
-            _generators.emplace(movement);
+            _generators.emplace_hint(insertPosition, movement);
```

Comment:
this will make sure the new generator is added in front of all others with the same priority